source/opt/function.cpp

```diff
@@ -193,6 +193,15 @@ BasicBlock* Function::InsertBasicBlockBefore(
   return nullptr;
 }
 
+bool Function::HasEarlyReturn() const {
+  for (auto& block : blocks_) {
+    if (spvOpcodeIsReturn(block->tail()->opcode()) && &*block != tail()) {
+      return true;
+    }
+  }
+  return false;
+}
```

Comment:
This takes a syntactic approach and identifies a return that is not in the last syntactic block.

Can we change it so that it identifies early returns regardless of block layout?

Here are some tests that illustrate what I mean:

[patch.txt](https://github.com/KhronosGroup/SPIRV-Tools/files/4918650/patch.txt)

And here is a proposed implementation of HasEarlyReturn that I think should give us what I think we need here:

```
bool Function::HasEarlyReturn() const {
  auto postdominator_analysis = blocks_.front()->GetLabel()->context()->GetPostDominatorAnalysis(this);
  for (auto& block : blocks_) {
    if (spvOpcodeIsReturn(block->tail()->opcode()) &&
        !postdominator_analysis->Dominates(block.get(), entry().get())) {
      return true;
    }
  }
  return false;
}
```

What this does is checks that if a block ends in return then it must be the case that every path from the entry block ends up at this block.  That captures what we want independently from syntactic layout.

Reply:
You're right. The block must have a return instruction and do not dominate the entry block.
Thanks!

Reply:
Done.